### ISSUE TITLE
level-5 extension: Error-handling, edge cases, and enum conventions

### ISSUE THREAD

**Issue by yusufaine** [nus-cs2103-AY2122S2/forum#21]
This is more of an "extension" of Level-5, and maybe enums.

I was cleaning up some of my code before starting on week 3 stuff and was thinking about the error-handling portion and had the following questions:

1. What are the edge cases you guys considered?
    * These are mine (on top of the given example):
      * out of bounds
      * no delimiters (`/at` or `/by`)
      * index not being a number
2. How did you guys organise the error-handling portion for yours?
3. Is there a convention for error-handling?
4. How do you _actually_ use enums?

Currently I'm just throwing `DukeExceptions` with the respective messages and tried to account for the edge cases but that doesn't seem "right".

I've also separated my "commands" into an enums class and did all the responses and error-checks there. 

I was thinking that this was the way as it would clean up a the main Duke and it _felt_ like enums was where all the responses/actions (mark/unmark/delete) _should_ happen. During the code clean-up, however, I realised that all I'm doing is just making it messier in my enums class (example below), and had me thinking about how to properly (conventionally) approach this.

```java
public void toggleMarkResponse(List<Task> taskList, String cmd, String toggleIdx) throws DukeException {
    if (taskList.isEmpty()) { throw new DukeException("The list is empty!"); }
    if (!argumentsCheck(toggleIdx)) { throw new DukeException("Invalid number of arguments!"); }

    // checks input to see if it's a number, throws DukeException if it can't be parsed
    int index = isValidInt(toggleIdx) - 1; 
    if (!isValidIndex(index, taskList.size())) { throw new DukeException("Entry out of range!"); }

    {... mark/unmark logic ...}

    {... all the System.out stuff ...}
}
```

**Comment by jbkim1999** [comment]
For me, when considering todo, event, and deadline command, I check these conditions:

Whether content of the task is specified. (an empty content, or delimiter follows immediately after the command)
Whether the schedule is specified (delimiter not specified at all or after delimiter no schedule is written)

For others, I only accept the "correct" conditions and all else will throw an exception.

**Comment by lkldev** [comment]
Hi, interesting question about enums! I was trying out using enums for the various commands and I decided to use two enums to help me handle the commands. The first enum `ActionType` would be to group the command into various types, such as `READ`, `UPDATE`, `ADD`, and `NO_ACTION`. The next enum `CommandAction` would be to create all the commands (`todo`, `list`, `deadline` etc) and tie them accordingly to `ActionType`. This enum will also contain information such as their required arguments etc. This way, I can easily use a hashmap to tie the input to the right command (and its arguments) and when handling exceptions, I can handle by their `ActionType` for some cases. For instance, the `UPDATE` `ActionType` commands would include `mark`, `unmark`, and `delete`, which all would require an index. When parsing the command, I can do something like:
`
switch(ActionType):
    ...
    case UPDATE:
        if (args.length < 2) {
            throw new IllegalArgumentException(String.format("The index of %s cannot be empty.", args[0]));
        }
     ...
` 
However, after looking at [se-edu/addressbook-level2](https://github.com/se-edu/addressbook-level2), I feel that enums may not be the best way to parse the commands, as creating the various commands as classes seem to allow for better polymorphism. Hope this help with your question regarding how enums can be used!

**Comment by Spyobird** [comment]
Hi, I don't think there is a right way to handle exceptions (although there are poorer ways ie. control flow). At the end of the day, whatever works best for you is good, although it is nice to keep OOP principles in mind. Coupled with the iterative breadth-first approach this module takes, I think it is also good practice to continually change the design of your code each week to meet the requirements. As for enums, use it whenever you want to! You could even use enums to differentiate the different exceptions/errors if you wanted to. I used enums in a very simple way to differentiate marked tasks and unmarked tasks, meanwhile I used polymorphism to differentiate exceptions. Hope it helps! :)

**Comment by yusufaine** [comment]
@jbkim1999 thanks for pointing out regarding the content of task and the schedule, that slipped my mind! (Looks like I'll have to add 2 more conditions 😆 )

---

@lkldev I think your approach might be the way, at least from a code-organisation perspective. I wanted to do something similar but opted not. I gave it some thought and realised that it might look cleaner, but that's adding another "layer" of passing arguments, which isn't necessarily bad, but I've had the bad habit of over-complicating simple tasks (which is also why I posted this question to see if what I had was over-engineered 🤣 ). 

> after looking at se-edu/addressbook-level2, I feel that enums may not be the best way to parse the commands

Agreed! But while polymorphism (and cleaner code) is off the table, overloading still can be used, though it is not as ideal, and I personally got very confused and resorted to making different exceptions/responses according to the respective enums.

I've actually done something similar to yours --
<details>
<summary> Click here to see the code </summary>

``` java
switch (inputCmd) {
    // ...
    case TODO:
    case EVENT:
    case DEADLINE:
        if (input.length == 2) { // here I've used .split(" ") to split the command from description
            // subtaskResponse checks the enum and creates the respective events 
            // + checks for delimiters, and list size
            inputCmd.subtaskResponse(taskList, input[1]); 
            break;
        } else {
            throw new DukeException("Task description cannot be empty!");
        }
    // ...
}
```

</details>

---

@Spyobird thanks for sharing your views on this too! This is my first time really building anything and I was worried that I might have done some "bad practices" regarding enums. I should probably take a look at how you (and @lkldev) used polymorphism to differentiate exceptions as it sounds cleaner 😄 .

---

I'll be leaving this open for a while more in case anyone wants to chime in, and will close it by the end of today if there are no further discussions.
